### DISCUSSION TITLE
How feasible would it be for a Redhat based port to be created?

### DISCUSSION THREAD

**Discussion by rkwesk** [ltsp/ltsp#197] [Q&A]
It is my opinion that many users of a Redhat based distro would benefit from this amazing work. I have reached out to someone who has written some documentation for Fedora on ltsp two or three years ago, telling him about this wonderful ltsp19!

Why not share such a gem with as many more people as possible was my thinking.

Richard

**Comment by enaut** [comment]
I would welcome that too.

I think so far it's just missing time, funding or volunteering to do a port.

To be included in the official Fedora repos there would need to be some style changes as ltsp19 does many things differently than the way Fedora advises to do them.

It might be better to fork ltsp or rather to redo the necessary steps in a rpm-distro-fashion.

Also ltsp is targeted mostly to schools. Teachers at schools typically use ubuntu - because "that is already hard enough".

just my opinion on the topic ;)

**Comment by alkisg** [comment]
> I think so far it's just missing time, funding or volunteering to do a port.

True; personally since neither me nor my schools use rpm-based distributions, I'd need to find some funding to work on that.

> To be included in the official Fedora repos there would need to be some style changes as ltsp19 does many things differently than the way Fedora advises to do them. It might be better to fork ltsp or rather to redo the necessary steps in a rpm-distro-fashion.

Any developers would be welcomed to join the LTSP team, with or without Fedora experience.
I believe the LTSP code can be adjusted to be compliant with all distribution policies. For example, I was careful while designing its architecture so that it would work both when sent over TFTP (ltsp.img) and when provided locally (initramfs-tools or dracut).
Forking a project with low development power isn't really a good approach, as it leads to less maintained forks.

Additionally, I think the new LTSP architecture results in a lot less distro-specific code, compared to LTSP5.

**Comment by rkwesk** [comment]
Yes, many choose Ubuntu for the reason you stated. In my opinion ltsp is worthwhile for many time constrained organizations (including but not only schools) and many of these already use a Redhat based system.

Alki, I also have the feeling that this latest code is less dependent on Debian based setups and I hope others elsewhere become aware and will recognize how useful and (relatively) easy it would be to work on such ports. Alas, my skills do not measure up.

Richard

**Comment by rkwesk** [comment]
I was searching the net for active ltsp people with rpm distros and came across this discussion in:

     https://fedoramagazine.org/how-to-build-a-netboot-server-part-1/

I know that a netboot server is not the same but some of the comments suggested that those people would/could contribute if they were made aware of ltsp19. I emailed the two whose name had links but often email from unknowns go straight in to spam.  :((

This weekend I will try Fedora 30 and simply try to compile the source file of ltsp19 and see what happens.

Richard

**Comment by alkisg** [comment]
LTSP19 is a netboot (fat client) server, not a terminal (thin client) server. The author of the blog you linked says he wants to do a part-2, with writable home; which is what LTSP already offers, along with lots more.
That probably means that the author there would love to have the new LTSP on Fedora.

Users testing LTSP on Fedora now won't be able to get it running, nor will they be able to give valuable feedback. A developer is needed:

The part of the Fedora port that will take the most time is dracut integration. After that is done, all initramfs-tools (.deb) and dracut (.rpm) based distributions will be very easy to support. Dracut does support the nfs=server:/path syntax that ipxe already has, so that we can have distro-agnostic command lines, but it does need the appropriate hooks to call the ltsp-bottom applet.

**Comment by rkwesk** [comment]
Alright. However, I did an install of Fedora 30 but cannot get epoptes. I thought you should be aware of this also.

     Running transaction
     Preparing        :                                                            1/1 
     Running scriptlet: epoptes-client-0.5.10-4.1.noarch                           1/1 
     /var/tmp/rpm-tmp.ctsTeR: line 1: fg: no job control
     error: %prein(epoptes-client-0.5.10-4.1.noarch) scriptlet failed, exit status 1

     Error in PREIN scriptlet in rpm package epoptes-client
     Verifying        : epoptes-client-0.5.10-4.1.noarch                           1/1 

     Failed:
     epoptes-client-0.5.10-4.1.noarch

     Problem: conflicting requests
     - nothing provides python-pycha needed by epoptes-0.5.10-4.1.noarch
     =======================================================================
     Package         Architecture   Version              Repository                Size
     =======================================================================
     Skipping packages with broken dependencies:
     epoptes         noarch         0.5.10-4.1           rpmsphere-noarch         258 k

     Transaction Summary
     =======================================================================
     Skip  1 Package

     Nothing to do.

Richard

**Comment by alkisg** [comment]
Richard I've never installed or tested Epoptes in Fedora; if you want to run it there, you should probably check who was the maintainer, and see if he's still interested in maintaining it.

Deb files do have a maintainer email address; maybe Fedora .rpm files also have that information.

It happens that Epoptes upstream developers are also Debian maintainers; but we don't maintain any other distributions; bug reports for them are to be filed downstream.

**Comment by enaut** [comment]
@rkwesk I don't see epoptes in the repository (F31) where did you get an rpm? I'd suggest to compile it from source instead.
The problem is that python-pycha has either been renamed or your rpm-file is for a different distribution having a package named that way.

**Comment by alkisg** [comment]
Epoptes 0.5.x was based on Python 2;
Epoptes 1.x is based on Python 3 and doesn't use pycha at all.

If anyone tries to get Epoptes running in a recent Fedora, I'd suggest they'd try with 1.x.

**Comment by kimheino** [comment]
I'll copy-paste my mailing list message here:

```
I've ported ltsp 5 to rhel 7 and using it on various production
environments. Source repositories are available at launchpad.

I'm also going to port ltsp 19 to rhel 8, but it will take some time.
Probably early next year? I will do that alone if needed, but of course
all the help is welcome.
```

Launchpad = https://code.launchpad.net/ltsp

**Comment by rkwesk** [comment]
The author responds:

Richard Kweskin <rkwesk@yahoo.com>
To:gbartho@siue.edu
Oct 24 at 5:05 PM
Hi Greg

I am not a part of the development team of ltsp here in Greece but have implemented their latest version called ltsp19 - 19 is for the year of conception - which is a total rewrite of ltsp (now called ltsp5.)

It works much easier with uefi and the code is much neater. I am hopeful that I can get a discussion going on an rpm port as the Greek team has concentrated on Debian's deb.

Have a look:

https://github.com/ltsp/


Richard Kweskin
retired IT



Bartholomew, Gregory <gbartho@siue.edu>
To:rkwesk@yahoo.com
Oct 28 at 2:35 PM
Hi Richard:

It is great to see that someone is reviving the LTSP! It is something that I would like to up on some of my systems at some point. Unfortunately, I'm a bit busy with some other projects right now.

An RPM for LTSP would be awesome! I'm not a package maintainer though. Have you seen Ankur's articles on maintaining RPMs?:
https://fedoramagazine.org/author/ankursinha/

Good luck with your project. Sorry that I don't have the time to help out right now.

Greg Bartholomew

**Comment by rkwesk** [comment]
Without knowing Fedora 31 was being released I installed version 30 and found nfs and dnsmasq already installed. I tried to install epoptes and found that once I installed an rpm called sphere it gave me both epoptes and epoptes-client. However, their versions were 0.5 and so I came aground with psycha.

Searching Fedora packages for in https://pkgs.org/download/epoptes did not list Fedora 31 or epoptes 1.0.

Richard

**Comment by alkisg** [comment]
Richard for Epoptes it would probably be better to use Epoptes issues, not LTSP issues.

Some more information regarding LTSP on RHEL.

The most difficult part would be the dracut (initramfs) integration. I've played with it a bit in the summer with partial success but I would need a couple of weeks of studying in order to things properly.
Someone with experience in dracut would do it much faster, but he'd have to spend some time to learn the new LTSP code base.

I'm willing to help such a person. If Kim's that person, I hope we'll be able to find some common free time to do this together.

After dracut integration is done, the rest of the things will be very easy. Maybe the only needed change will be in PAM integration (a couple of sed commands) and nothing more.

**Comment by alkisg** [comment]
Btw here are some dracut bugs that I filed (or participated in) while I was testing it:

- https://github.com/dracutdevs/dracut/issues/380
- https://github.com/dracutdevs/dracut/issues/433#issuecomment-463233662
- https://github.com/dracutdevs/dracut/issues/539
- https://github.com/dracutdevs/dracut/issues/540

We don't care about nbd anymore, but we would really prefer to use the stock nfsroot kernel option.

**Comment by alkisg** [comment]
I managed to netboot fedora33 with LTSP and autologin to a graphical session, with the following workarounds and restrictions:

- Inside the chroot, I added nfs support in the initrd with: `dracut --no-hostonly --force --add "nfs squash" --add-drivers "nbd squashfs OVERLAY"`
- In ltsp.ipxe, I added: `root=/dev/nfs nfsroot=${srv}:/srv/ltsp/${img},vers=3 init=/usr/share/ltsp/client/init/init`
- This means that no LTSP code went into the dracut initramfs, and 25-ro-root.sh was used to apply the writable overlay, as it's done on Raspberry Pi OS
- TMPDIR was needed in ltsp/ltsp: `TMPDIR=/run locate_applet_scripts "ltsp"`
- 54-pam.sh doesn't work in Fedora yet, and selinux was causing some problems, so I had to add these in ltsp.conf:

```shell
[clients]
AUTOLOGIN="administrator"
FSTAB_HOME="server:/home /home nfs defaults,nolock 0 0"
PAM_AUTH_TYPE=0
POST_INIT_SELINUX="sed 's/SELINUX=enforcing/SELINUX=disabled/' -i /etc/selinux/config"
```

**Comment by SchoolGuy** [comment]
Okay since this is moving forward on a regular basis, I would love to hear what needs to be done to support this on openSUSE/SLE based distros? My laptop sucks and I have a fairly powerful server left at home, thus I would be willing to maintain the RPM from an openSUSE side, if I would get some help with the inital work.

As a second step i would propose to make this also work on transactional based distros like MicroOS because that would make the server then literally indestructible and auto-updating.

**Reply by SchoolGuy** [reply]
I will try to have a look around myself some time soon in a VM.

**Comment by alkisg** [comment]
Hi Enno, thanks, I'm sure many people would appreciate an LTSP RPM maintainer.
The server part should be rather easy, you can start with that if you want, i.e. to allow someone to `yum install ltsp dnsmasq nfs-server ...` and follow the ltsp.org/docs/installation instructions, and be able to have an openSUSE server with e.g. a Debian client image.

The client part is a bit more difficult. You should be able to netboot openSUSE by applying the workaround that I mentioned in my previous message, but for a real port, the dracut and pam integrations should be implemented. For these, I think you should first read the code and see it in action with a debian-based distribution, so that you can see which things need to be ported in dracut/pam.

I'll be around to give as much advice as I can, but at this time I won't have the time to offer dracut-related code.

**Reply by SchoolGuy** [reply]
Thanks a lot. I will try to get started soon. I will post the link to my home project in OBS as soon as I have it created.

**Reply by enaut** [reply]
If you make the progress visible (a fork or something) I might help! That is if I find the time. But I'd really appreciate a Fedora based ltsp.

**Reply by SchoolGuy** [reply]
@enaut The plan is to package it and have the distro specific patches in the OBS until upstream has released a new version where they are then integrated (obviously I will open PRs for that), I definitely won't have the time to maintain a complete fork of such a project.

**Comment by alkisg** [comment]
I think a public fork e.g. in https://github.com/SchoolGuy/ltsp is a good idea.
Multiple persons can collaborate in such a fork.
A pull request can be filed, when it's ready to be merged.
After it's merged, it can continue to be maintained upstream in LTSP.

I would like the deb/rpm packages for all supported distributions to be generated with the same method.
That means that the current PPAs will need to be replaced.
OBS is a good choice.
Since the ltsp package doesn't need to be compiled, it might also be possible to do everything directly on github, i.e. to build the packages using github actions and to create and host the repositories on github. But that will need a lot of research, so OBS is probably the way to go.

**Comment by SchoolGuy** [comment]
I will start working here soon (TM): https://build.opensuse.org/project/show/home:SchoolGuy:ltsp

**Reply by svetoi1985** [reply]
Hello, can you describe in more detail how you did it. Thank you in advance!!!
_____________________________________________________________________________________________________________________________________________

**I managed to netboot fedora33 with LTSP and autologin to a graphical session, with the following workarounds and restrictions:

Inside the chroot, I added nfs support in the initrd with: dracut --no-hostonly --force --add "nfs squash" --add-drivers "nbd squashfs OVERLAY"
In ltsp.ipxe, I added: root=/dev/nfs nfsroot=${srv}:/srv/ltsp/${img},vers=3 init=/usr/share/ltsp/client/init/init
This means that no LTSP code went into the dracut initramfs, and 25-ro-root.sh was used to apply the writable overlay, as it's done on Raspberry Pi OS
TMPDIR was needed in ltsp/ltsp: TMPDIR=/run locate_applet_scripts "ltsp"
54-pam.sh doesn't work in Fedora yet, and selinux was causing some problems, so I had to add these in ltsp.conf:**

**Comment by alkisg** [comment]
@svetoi1985 as no developer is interested on maintaining the RHEL portion of LTSP yet, and no users are interested in sponsoring such a port, there's no point in wasting time on this...
Sorry my time for working on open source projects got very limited during the last years!
